### DISCUSSION TITLE
Dexcom G6: Collector must be continuously restarted to get data

### DISCUSSION THREAD

**Discussion by tubenn** [NightscoutFoundation/xDrip#2204] [Q&A]
Dear Xdrip+ comm,

First of all, thank you for an amazing app and community - truly a life changer for me.

Following an update to the latest OS version on my OnePlus 9T (not the ideal choice for Xdrip+ I now realize, given the agressive app management), I am faced with the issue that approximately every second reading from my Dexcom G6 is missing. It seems that the Bluetooth link gets stuck in Disconnecting or Connected, and is saved by the great "agressively restart collector" feature (also works when I manually trigger a collector Restart). However, the missing data is interfering with my closed loop. I have tried a number of settings i. Xdrip+, reset battery optimization etc, but not been able to crack this issue yet.
Anyone who have experienced similar issues and solved it, short of buying a new phone? I was using the stable version but have now updated to the nightly build without effect (currently 6171850-2022-07-01). Else, a feature request could be to force a collector Restart every 4.5 min.

Thanks a ton,
Tue

**Comment by Navid200** [comment]
Can you post a screenshot of the following page?
Settings &#8722;> G5/G6 Debug Settings

Can you also post a screenshot of the system status page?  You can blackout the transmitter ID if you like.

**Reply by Miquest** [reply]
I have the same problem. It doesn't help to change the Settings, it will bee only more worse than before.
I agree with the Suggestion of @tubenn,
but the Funktion needs to use Wakelocks every 2 minutes to prevent killing by the operating system. That's my experience with that. If I use the phone continuously, it's working perfectly. For example tracking an Activity with Adidas Running. We need System Wake Locks to solve this problem I think.

**Reply by Navid200** [reply]
There is no debate that some phones have intermittent connectivity problems.
But, there are many more phones that don't.
xDrip (the collector) is improved to address confirmed issues.

The only way to confirm that a particular phone needs attention is to demonstrate it misbehaves with the recommended settings.
It is so easy to make any phone have problems using a single incorrect setting.

**Reply by tubenn** [reply]
Thank you for the prompt replies @Navid200 and @thorinEichenschild007 👍

Here follows the screenshots. On the status page is an example where it is stuck on disconnecting. It will miss the next reading at 5 min, but then it appears that the service restarts (or a manual Restart service) and the reading at 10 min is picked up. Any inputs will be greatly appreciated. Again, if this is an issue that is affecting a significant number of people, my guess would be that a solution could be to enable the aggressive service Restart every 4th min or so.

Thanks!

![Screenshot_2022-07-02-22-09-39-56_2dd2c1bcb8f659cd696bcba2072f751b~2.jpg](https://user-images.githubusercontent.com/40670391/177015431-402a0f00-16a4-4115-b790-9e4027701bc1.jpg)![Screenshot_2022-07-02-22-04-23-30.jpg](https://user-images.githubusercontent.com/40670391/177015434-96af94cf-ba7c-496b-a4dc-4745dcbbc5ee.jpg)

**Reply by Miquest** [reply]
You have a OnePlus or Oppo Phone, right? I have one too...
Since the Upgrade to OxygenOS 12 I have the same Issue with the long WakeUp Time...

**Reply by Navid200** [reply]
Is there a reason you have engineering mode enabled?
I would disable that.

And I would change my settings to match the following:
https://navid200.github.io/xDrip/docs/G6-Recommended-Settings.html

After you do all of that, please let me know if the problem persists.

**Reply by tubenn** [reply]
Thank you for the detailed instructions. I made all changes, part from the Preemptive restarts which I am dependent on. I use a rechargeable G6 transmitter which I hard reset every 2 months, hence the engineering mode. I have two transmitters, and the behavior is identical.

After making the changes, the issue persist. After a reading the BT link status is Disconnecting and the reading after 5 min is missed. I now see that after 6 min there is a power connect which at 7 min ends with BT link status changing to Disconnected. The reading at 10 min is then received. Manually pressing Restart collector every ~4th min fixes the issue. Of note, I kept my OnePlus 9T screen on throughout the test.

![Screenshot_2022-07-02-23-16-44-82_2dd2c1bcb8f659cd696bcba2072f751b~2.jpg](https://user-images.githubusercontent.com/40670391/177016622-0970afea-7d79-4835-89c5-bb47e7929697.jpg)
![Screenshot_2022-07-02-23-18-14-65_2dd2c1bcb8f659cd696bcba2072f751b~2.jpg](https://user-images.githubusercontent.com/40670391/177016624-c0f4cb02-a47f-4dda-b4cf-adcabf21f945.jpg)
![Screenshot_2022-07-02-23-18-52-26_2dd2c1bcb8f659cd696bcba2072f751b~2.jpg](https://user-images.githubusercontent.com/40670391/177016625-b01e53e6-7701-4d85-9982-273265a6dbfb.jpg)

**Reply by Navid200** [reply]
Please disable engineering mode.
There has been no need to enable engineering mode in order to hard reset a Dexcom transmitter since August 2018.

**Reply by tubenn** [reply]
Good to know @Navid200 , I disabled it when I followed your instructions. The last message and screenshots were therefore without engineering mode.

**Reply by Miquest** [reply]
I'm also without engineering mode...
I reinstalled xDrip Plus totally and installed it new with the recommended Settings, but it doesn't help...

**Reply by Navid200** [reply]
@tubenn Can you show a screenshot of the BG data table?

**Reply by tubenn** [reply]
@Navid200 Certainly. Sorry, but do you mean the Error and Event list (below here)?

![Screenshot_2022-07-03-14-34-28-99.jpg](https://user-images.githubusercontent.com/40670391/177039921-cc81c7b9-ab00-4252-be08-2840ff0cdb23.jpg)

**Reply by Navid200** [reply]
No, I have explained the data tables here: https://navid200.github.io/xDrip/docs/Datatables.html

But, I don't understand what could be causing the backfill timestamp to jump so far back to the past and then to the future as your logs show.  Do you see those logs often?

**Reply by tubenn** [reply]
@Navid200 Excellent. Data tables comes here.

These errors/earnings only appeared after changing to the default settings (at least I haven't noticed them before)

![Screenshot_2022-07-03-14-56-50-53.jpg](https://user-images.githubusercontent.com/40670391/177040897-8fc2222a-d12f-4262-a05c-e9319197195e.jpg)

**Reply by Navid200** [reply]
Can I see your preemptive restart parameters?

**Reply by tubenn** [reply]
Absolutely

![Screenshot_2022-07-03-15-14-05-41_2dd2c1bcb8f659cd696bcba2072f751b.jpg](https://user-images.githubusercontent.com/40670391/177041406-da333dc0-67fe-4db1-9cca-9d19b7fece30.jpg)

**Reply by Navid200** [reply]
Looking at the data table, I see one native followed by two backfills and that sequence continues.  If you scroll back, does that pattern continue?
Or, do you ever see two consecutive natives with no backfill beside it?

**Reply by tubenn** [reply]
That is the pattern. I can force a double native read without backfill using Restart collector after the first read. Of note, prior to reverting to the default settings, it was only every second reading that was backfilled.

**Reply by tubenn** [reply]
![Screenshot_2022-07-03-15-20-37-16_2dd2c1bcb8f659cd696bcba2072f751b.jpg](https://user-images.githubusercontent.com/40670391/177041673-5420e15b-d59f-4466-bca3-e44ec4d49582.jpg)

**Reply by Navid200** [reply]
I need to ask for advice.  Please be patient.

**Reply by tubenn** [reply]
Regardless of the outcome, thanks a ton @Navid200 for your time and help!

**Reply by tubenn** [reply]
Slight development, not sure if you can use this info but here goes. I realized that no data had been downloaded for nearly an hour. After this. I had to manually restart the collector prior to any reading. Disabling "Avoid scanning" fixed this issue and brought me back to getting every third reading (i.e. successfully connecting every 15 min to the G6 transmitter, so 1x native reding followed by 2x backfill every time). Disabling "Minimize Scanning" further brought this down to every second reading (i.e. successfully connecting every 10 min and getting 1x native anz 1x backfill data).
As mentioned, these issues started after an OnePlus Oxygen update, so it is a phone-specific issue, likely related to battery optimization.

**Reply by Miquest** [reply]
I think so too...
But I deactivated the Doze Mode completely via ADB, it didn't help. 
We need a wakelock or something like that...

**Reply by Miquest** [reply]
Buggy OnePlus Workaround...
Maybe that's also helpful for other Smartphones...
The workaround should include the restart of the OB1 Collector every 2minutes.
Or a wakelock.
That's what a Workaround should do.
Solving a problem with other methods as usual.
And that methods are continuous wakelocks and a restart of the collector.
The manual Workaround is working. But we need something which can do that what we do on the software side.
Restarting and waking up

**Reply by tubenn** [reply]
@thorinEichenschild007 I like the feature-name suggestion.

**Reply by Miquest** [reply]
Do you agree with my opinion? I think that's exactly what we need...

**Reply by jamorham** [reply]
I'm pretty sure the problem here is the slow wakeup which is where the phone violates the android framework standard by not waking up when we ask it to. Almost always this will be due to manufacturer power-saving features. These can take many different forms and the whitelisting can be hidden deep in various menus. Sometimes it is referred to as background execution. Assuming your phone has a way to disable power-saving like this on an app-by-app basis then you need to dig in and try to find it. Sometimes there are multiple things that need to be switched off.  If not reviewed already, take a look at this: https://dontkillmyapp.com/oneplus and see if those options resolve the problem. If one does please let us know which setting(s) helped.

**Reply by Miquest** [reply]
I read this instruction several times but no option helps.  I also think the battery optimization is the problem.  However, I also completely deactivated the Doze Mode via ADB, but that also had no effect on the WakeUp Time.

**Reply by Miquest** [reply]
![Screenshot_2022-07-04-10-39-00-03_40deb401b9ffe8e1df2f1cc5ba480b12](https://user-images.githubusercontent.com/91971937/177116762-0ca3805e-f422-422c-b108-ee34d3397a85.jpg)


On dontkillmyapp.org you can read that you cannot configure the Battery Optimization like we need to prevent xDrip from App Killing...

**Reply by jamorham** [reply]
The whole point of foreground services is they don't get killed so that seems really bad if so. ~~Your only option in that case would be to use the `trust auto connect` xDrip setting which would also might not work if services are being killed and not just suspended.~~ (Please disregard that last suggestion as I don't think that setting differentiation is present in the mainline release) There may be other options if the phone is rooted. Also down-grading the phone might be another option but also very disruptive.

**Reply by Miquest** [reply]
In case of rooting, what would you do?
A custom ROM?
"Trust Auto Connection" is enabled.

**Reply by jamorham** [reply]
With rooting, if xDrip was placed in the `priv-app` folder as a system app then it is quite possible that the phone will be much less likely to prevent its background processing. Otherwise changes would be made to the rom, for example disabling the aggressive power management. I am unfamiliar with how that would be achieved on this phone but those are the types of steps which would be taken to resolve the issue if there is no way in the phone settings or in xDrip without rooting.

**Reply by Miquest** [reply]
Thank you so much for the quick reply @jamorham .

I will try that tomorrow and I'll let you know if it works!

**Comment by Navid200** [comment]
@tubenn Have you disabled battery optimization for Bluetooth also?

Are all your xDrip Bluetooth settings in agreement with the suggestions?
Do you have "Aggressive service restarts" enabled in Settings &#8722;> Less common settings?

**Reply by Miquest** [reply]
Disabling Battery Optimization for Bluetooth is not possible on OnePlus Phones!
Maybe if you root it, but in the normal system Settings that's not working.

**Reply by tubenn** [reply]
@Navid200 All settings are as recommend, including aggressive service Restart. I suspect that this setting is what saves me, and my initial idea was if this feature could be made more agressive, hence triggered before every read per default.
I followed your nice and clear guides, and I have also setup Xdrip as recommend for app OnePlus with enabled background activity, auto launch, locked app, etc.

However, like @thorinEichenschild007 writes it is not possible to change BT setting, and after the latest Oxygen OS update and some settings like advanced battery optimization have been removed, so I do not know if they are on per default now, which would be bad.

**Reply by Navid200** [reply]
There are two factors that make improvements to the collector extremely challenging.

The first is that the developer needs to have a sample of the problematic phone.  This is costly.

The second is that modifying the collector will affect all phones.  Fixing one and breaking another in the process is a possibility that makes it difficult.

jamorham has been doing this, fixing connectivity problems, for years.  Please don't think this is the first time something like this has happened.

There have even been individuals who have gone as far as opening a PR.  This is an example:
https://github.com/NightscoutFoundation/xDrip/pull/2121

**Reply by Miquest** [reply]
I think the complete development team is a genius group of persons, xDrip is so much better than the original Apps for Dexcom, Libre...

But we have no problem with unwanted unbonding like in the post you linked, we have a problem with battery optimization.
I'm really interested to test some new features only for developers, I could be a testing person for that. But I think that a Wakelock couldn't crash the complete system...
Or a restart of the Collector.
That's exactly what @tubenn and I do manually.
I tried Automagic Premium to fix that, but that was also killed by the Operating system after 20 minutes after I installed it, and then every time.

**Reply by tubenn** [reply]
@Navid200 Please don't read my comments as a blame-game. I have vast respect and am grateful for the devs and the freetime they spend on this helping us all. I am also mindful of the butterfly effect when changing code, hence my original post was more getting advice on if anyone found working settings for this new issue on the most recent OnePlus OS. Next time I will likely buy another phone brand, but xDrip+ is staying. So thanks to you and the devs @Navid200 ! I really appreciate your efforts so far!

**Comment by tubenn** [comment]
I installed the most recent nightly build and Xdrip now states "Buggy handset: Using workaround" which seems to fix the issue as data now is received.

Thanks a ton to the devs!
